### ISSUE TITLE
`Plagiarism`, `Programming Exercises`: Broken code view

### ISSUE THREAD

**Issue by b-fein** [ls1intum/Artemis#6579]
### Describe the bug

In the plagiarism view for programming exercises the file view on the right side remains empty. 

### To Reproduce

1. Have a programming exercise with a few submissions.
2. Run the plagiarism check.
3. Select a case marked as potential plagiarism.
4. The file for student1 on the left side loads, the one on the right side does not.
   - See log output in your browser console as attached at the end of this issue.
5. Changing to different files does not change that.

### Secondary bug, same underlying issue

1. Steps 1-3 from above.
2. Change to another student pair on the left side.
3. Try selecting different files in either left/right view.
4. The shown file contents now no longer match the selected names at the top. (Most of the time, for some files it seems to work.)
5. The same console error appears.

### Expected behaviour

Both files are shown.

### Which version of Artemis are you seeing the problem on?

6.1.5

### What browsers are you seeing the problem on?

Firefox

### Relevant log output

```shell
ERROR TypeError: c[D] is undefined
    insertMatchTokens text-submission-viewer.component.ts:244
    next text-submission-viewer.component.ts:178
    RxJS 29
    Angular 17
    RxJS 64
    handleFileSelect text-submission-viewer.component.ts:175
    RxJS 16
core.mjs:8416:22
```

**Comment by jakubriegel** [comment]
Hi! I've verified on `6.1.7` and the issue doesn't happen there. Can you check on `6.1.7`?

**Comment by b-fein** [comment]
Thanks for checking.

I can’t replicate it with the exercises on TS2 at the moment either. Will retest on our production system that has enough data for that as soon as we upgrade.
According to the changelog there aren’t any changes that would be relevant to that view, though, so I’m not sure what’s going on there.

**Comment by jakubriegel** [comment]
Ok, let me know what's the result

**Comment by krusche** [comment]
There might have been an issue related to a dependency update which made apparent that come client code used injected services before the constructor was invoked which leads to rendering issues which was fixed

**Comment by krusche** [comment]
But it can also be related to specific submission data

**Comment by b-fein** [comment]
(Artemis 6.2.0)

Now files are shown, but the first match in each file is repeated an additional nine times:

![image](https://github.com/ls1intum/Artemis/assets/64250573/5af36d20-aa2f-423d-940a-0b57773b8bbd)

Same console error:
```
ERROR TypeError: u[z] is undefined
    insertMatchTokens text-submission-viewer.component.ts:244
    next text-submission-viewer.component.ts:178
    RxJS 32
    Angular 9
core.mjs:10162:22
    handleError Angular
    handleError sentry.error-handler.ts:66
    next Angular
    RxJS 6
    Angular 11
```

**Comment by krusche** [comment]
ok, this is weird, we should fix that asap

**Comment by dearjasmina** [comment]
<img width="1704" alt="image" src="https://github.com/ls1intum/Artemis/assets/24391701/1e237f2f-6498-4453-8dba-318af6e4e359">
Here is again the same issue: the text was duplicated 2000 times :,)

**Comment by krusche** [comment]
@jakubriegel can you fix the issue please?

**Comment by b-fein** [comment]
I was debugging some of this, but cannot cannot really make sense of what the code is supposed to do:

### Outline


#### Crucial point 1: Question Q1

https://github.com/ls1intum/Artemis/blob/d76fc4171b93faddf61b7383fb910d5ca57f7948/src/main/webapp/app/exercises/shared/plagiarism/plagiarism-split-view/plagiarism-split-view.component.ts#L137

The `submission.elements` is a list directly from the server with as far as I can see no specific constraint on the order. How does the match start and length relate to the array index? Why should I be able to retrieve the pairs via those indices?


#### Crucial point 2: Assumptions A1, A2

##### A1: Different submissions

(A1) Since it is named a `FromToElement` I assume this is supposed to map a marked section in `File_1` in submission $A$ to the corresponding `file_2` in submission $B$.

Those pairs of matches are then used in the display component to highlight some parts:
https://github.com/ls1intum/Artemis/blob/d76fc4171b93faddf61b7383fb910d5ca57f7948/src/main/webapp/app/exercises/shared/plagiarism/plagiarism-split-view/text-submission-viewer/text-submission-viewer.component.ts#L228
In the linked code, the sections that should be highlighted in `File_1` are computed by taking the starting line and length into account.
If my assumption (A1) is correct, then `match.to.line` should be the starting line of the match in `file_2` of submission $B$. In that case it makes no sense to use that as the end marker for the highlight in `file_1` of submission $A$.


##### A2: Same submission

(A2) Since it is named a `FromToElement` I assume this is supposed to map the start of the marked section in `file_1` in submission $A$ to the end of the marked section also in `file_1` in submission $A$.

In this case: Why would both items have a length? Why is this not represented as a single `{ from: i, to: j, col_from: k, col_to: l }` object?


### Example

`match.from`:
```js
{ column: 5, line: 137, file: "src/sudoku/model/board/Board.java", type: 0, length: 1 }
```
`match.to`
```js
{ column: 9, line: 246, file: "src/sudoku/model/board/SudokuBoard.java", type: 0, length: 38 }
```

The code now tries to build a highlighted section in `src/sudoku/model/board/Board.java` that spans from line $137$ to line $246$. Since this file only has $158$ lines, though, this results in the line access (`row[idxLineTo]`) returning `undefined` and this causing the error as in the initial issue description.

Assuming the list order as outlined in (Q1) is correct, this supports (A1). In case the list order (Q1) does not hold, no conclusion about (A1) over (A2) can be made since the pairs could be arbitrary.


### Conclusion

As far as I can tell, there are now two main (non-exclusive) possibilties how this could/should have worked:

- The list should indeed be ordered in some specific way (Q1), but no longer is for some reason. Therefore, the wrong items are matched and the `from`/`to` pairs are wrong.
- My assumtion (A1) is incorrect and the `FromToElement` represents something different like for example (A2).

:exclamation: Main takeaway:
In any case, the current plagiarism view for programming exercises seems to show pretty much arbitrary highlights even in cases where the file lengths happen to match and the component does not crash with the array-out-of-bounds access.

---

Additional remarks:

There are also `match.to` elements like
```js
{ column: 33, line: 71, file: "src/sudoku/shell/Shell.java", type: 0, length: -2118 }
```
I have no idea what the length is intended to mean in this case. → see (A2)

**Comment by coolchock** [comment]
Tested on ts4 with test user 15. The plagiarism matches are not duplicated any more, however the highlighting of the matches doesn't seem to work properly:
<img width="2560" alt="image" src="https://github.com/ls1intum/Artemis/assets/73854755/7346bb0a-834f-4d09-aa7e-26447295de5c">

I would expect that if the plagiarism is 100% both texts are blue.

**Comment by jakubriegel** [comment]
I worked on this several days already, but although I've seen this issue while using Artemis, I'm still not able to reproduce it on purpose

**Comment by jakubriegel** [comment]
The case reported by @egekurt123 and @dearjasmina still occurs for some submissions posted that time on ts4. In particular for test user 1 and 2. They are exactly the same and resulted in 100% similarity, but their contents are duplicated in the side by side view many times:
<img width="1512" alt="image" src="https://github.com/ls1intum/Artemis/assets/32958017/39f5dc35-8b36-49cb-8b1b-6ab6d3fd93f9">

Re-running checks does not fix the issue.

However, when I pasted the same submission in my local set up, all works fine:
<img width="1512" alt="image" src="https://github.com/ls1intum/Artemis/assets/32958017/5dd40a95-8973-46dc-b344-cfacaea8c8b7">

**Comment by krusche** [comment]
> I worked on this several days already, but although I've seen this issue while using Artemis, I'm still not able to reproduce it on purpose

Did you see @b-fein's analysis above? I guess you have to work with real data and you cannot simply use lorem ipsum.

**Comment by jakubriegel** [comment]
> > I worked on this several days already, but although I've seen this issue while using Artemis, I'm still not able to reproduce it on purpose
> 
> Did you see @b-fein's analysis above? I guess you have to work with real data and you cannot simply use lorem ipsum.

Yes, I went thorough it. These observations have more to do with the highlighting issue rather than repeated code issue I tried to reproduce. Fortunately, repeated code is now fixed by Benedict in #7841.

I already implemented the fix for code highlighting, but there are some important findings to share. I'll describe them in the next post.

**Comment by jakubriegel** [comment]
In #7854 I fixed the client, so that it properly interprets the matches coming from JPlag and visualises them in the viewer:
<img width="1273" alt="image" src="https://github.com/ls1intum/Artemis/assets/32958017/5ac39f28-fa22-470c-bb29-a6b941f9ae26">

The feature now works fine for all possible combinations of matches, but testing revealed an issue with JPlag matches for some programming exercises. Taking those two submissions as an example:
<img width="1307" alt="image" src="https://github.com/ls1intum/Artemis/assets/32958017/2e00ec64-5b04-4bdf-9297-a1d1ef29b791">

We can clearly see that highlighted parts of code do not match with the repeated/not repeated code. I manually analysed matches coming from the server and directly from JPlag. Findings are that the client properly visualises what was produced by JPlag and that JPlag produced invalid matches. To verify it we can check the raw JPlag report produced by Artemis:
<img width="1512" alt="image" src="https://github.com/ls1intum/Artemis/assets/32958017/af60c590-ae11-4e81-aa27-8ad81df39fff">

Highlighting here is a bit different from Artemis, but also not accurate.

@FuchsDominik Could you tell if this is an expected JPlag behaviour?

@krusche Should I continue with testing and reviewing #7854 as is?

**Comment by b-fein** [comment]
Can you please explain how your changes in #7854 relate to the found issue with the mismatching from/to-pairs in the example in my comment above?
As far as I can tell the from/to pair selection has not been changed.

**Comment by b-fein** [comment]
> We can clearly see that highlighted parts of code do not match with the repeated/not repeated code.

JPlag does not only check for directly equal text content. It also does some further analyses on the AST to detect other types of clones. So I think matching of non-text-identical parts is to be expected.

(you probably wanted to tag @dfuchss)

**Comment by dfuchss** [comment]
> > We can clearly see that highlighted parts of code do not match with the repeated/not repeated code.
> 
> JPlag does not only check for directly equal text content. It also does some further analyses on the AST to detect other types of clones. So I think matching of non-text-identical parts is to be expected.
> 
> (you probably wanted to tag @dfuchss)

@jakubriegel as @b-fein already explained, JPlag is not looking for equal text content. Thus, it's expected to find non-text-identical parts :)

**Comment by jakubriegel** [comment]
@b-fein @dfuchss Thanks for your input about JPlag!

**Comment by b-fein** [comment]
Thanks for the explanation!

> The FromToElement represents the start and end index of a single match in a single file. […] Matches are always in the correct order, thanks to the sort operation from lines 108 and 109 from the same file.

Okay, but then something else is wrong since the example from my comment is an actual one from a production exercise. `match.to` and `match.from` contain different file names.

**Comment by jakubriegel** [comment]
> Thanks for the explanation!
> 
> > The FromToElement represents the start and end index of a single match in a single file. […] Matches are always in the correct order, thanks to the sort operation from lines 108 and 109 from the same file.
> 
> Okay, but then something else is wrong, since the example from my comment is an actual one from a production exercise. `match.to` and `match.from` contain different file names.

The list of elements contains elements from all files involved in the submission, but a single match should reference only one file. Most likely the issue with indices and undefined elements also caused a shift here and accidentally Artemis used elements from two different files.

**Comment by b-fein** [comment]
Ran a fresh plagiarism check to make sure no old data remained.

Still broken in #7854 (lines might not match up with the code from the PR since I added some debug logging):
```
src/sudoku/model/EnforcedNumber.java (84): from: 58, to: 200
Object { id: 733900, column: 5, line: 59, file: "src/sudoku/model/EnforcedNumber.java", type: 0, length: 1 }
Object { id: 733746, column: 21, line: 201, file: "src/sudoku/shell/Shell.java", type: 0, length: -6996 }
ERROR TypeError: fileLines[idxLineTo] is undefined
    buildFileContentPartForMatch text-submission-viewer.component.ts:289
    buildFileContentWithHighlightedMatches text-submission-viewer.component.ts:243
    insertMatchTokens text-submission-viewer.component.ts:220
    next text-submission-viewer.component.ts:180
    RxJS
```

---

> Most likely the issue with indices and undefined elements also caused a shift here and accidentally Artemis used elements from two different files.

I’m not sure if that can happen:
1. The match list is sorted.
2. The from/to pairs for the matches are selected.
3. The from/to pairs are added to the text view, only then the `undefined` issue appears.
   - The component tries to access line `idxLineTo` (200) in the array of rows (`fileLines`) which only has 84 elements. 
   ⇒ The from/to pairing used to compute the section which should be highlighted already is wrong at this point (see log above), i.e., the wrong two items in the list are selected.

**Comment by jakubriegel** [comment]
@b-fein Could you generate in Artemis a JPlag report for that and open it on JPlag viewer? Maybe something is not right on the server or in JPlag

**Comment by jakubriegel** [comment]
In a separate conversation with @b-fein we discussed the raw JPlag report and that it renders correctly in JPlag viewer.

Unfortunately, I is not possible to work here on the submission that caused the issue (due to legal constraints). We also cannot provide any other example that causes similar issue.

What I was able to receive was the contents of the `GET /for-split-view` operation, which contains information necessary for the client to build the highlights: [for_split_view (1).json](https://github.com/ls1intum/Artemis/files/13894185/for_split_view.1.json)

Findings are as follows:
- When I manually inserted this JSON into the logic that calculates matches locations in the Artemis client, I realised that matches to and from indexes are calculated properly based on what's present in the `matches` list.
- Elements in both submissions objects contain strange token with negative lengths.
- JPlag assigns elements consecutive ids, but in the received response array elements are not ordered by those ids.
- I added a change in #7854, which always orders the list by id. Tests with mocked JSON showed that this removes matches from two different files (and is consistent with previous Benedikt's findings).
- @b-fein will test the updated code agains the broken submission and post his results.

**Comment by jakubriegel** [comment]
The `undefined` issue was caused by invalid ordering of submission elements in the server response. I fixed it by introducing client-side sorting of them. The fix can be found in #7854.

The same PR contains two other improvements: 
- fix of rendering of highlighting for text exercises
- introduction of full line highlighting for lines with matches in programming exercises, so that ambiguous matches are not as eye catching (with @krusche we concluded that this is the best option for now)

All three changes are described in the PR notes.

Thanks @b-fein for reporting this issue and for good cooperation!
